src/metax_api/services/file_service.py

```diff
@@ -779,64 +798,57 @@ def _get_directory_contents(cls, directory_id, recursive=False, max_depth=1, dep
         # to successfully also include files which did not DIRECTLY contain any files, but do
         # contain files further down the tree.
-        #
         # dirs which otherwise contained files, but not any files that were selected for
         # the cr, are not returned.
+
         sql_select_dirs_for_cr = """
             select d.id
             from metax_api_directory d
-            where d.parent_directory_id = %s
+            where d.parent_directory_id = {}
             and exists(
                 select 1
                 from metax_api_file f
                 inner join metax_api_catalogrecord_files cr_f on cr_f.file_id = f.id
                 where f.file_path like (d.directory_path || '/%%')
-                and cr_f.catalogrecord_id = %s
+                and cr_f.catalogrecord_id {} {}
                 and f.removed = false
                 and f.active = true
             )
             """
 
-        # select files which are contained by the directory, and which
-        # belong to the cr <-> files m2m relation table
-        sql_select_files_for_cr = """
-            select f.id
-            from metax_api_file f
-            inner join metax_api_catalogrecord_files cr_f on cr_f.file_id = f.id
-            where f.parent_directory_id = %s
-            and cr_f.catalogrecord_id = %s
-            and f.removed = false
-            and f.active = true
-            """
-
         with connection.cursor() as cr:
-            cr.execute(sql_select_dirs_for_cr, [directory_id, cr_id])
-            directory_ids = [ row[0] for row in cr.fetchall() ]
-
             if dirs_only:
-                file_ids = []
+                files = []
             else:
-                cr.execute(sql_select_files_for_cr, [directory_id, cr_id])
-                file_ids = [ row[0] for row in cr.fetchall() ]
+                if cr_id:
+                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '=', cr_id))
+                    files = None if dirs_only else File.objects \
+                        .filter(record__pk=cr_id, parent_directory=directory_id, removed=False, active=True) \
+                        .values(*file_fields)
+                if not_cr_id:
+                    cr.execute(sql_select_dirs_for_cr.format(directory_id, '!=', not_cr_id))
+                    files = None if dirs_only else File.objects.exclude(record__pk=not_cr_id) \
+                        .filter(parent_directory=directory_id, removed=False, active=True) \
+                        .values(*file_fields)
+            directory_ids = [ row[0] for row in cr.fetchall() ]
 
-        if not directory_ids and not file_ids:
+        if not directory_ids and not files:
             # for this specific version of the record, the requested directory either
             # didnt exist, or it was not selected
             raise Http404
 
         dirs = Directory.objects.filter(id__in=directory_ids).values(*directory_fields)
```

Comment:
Since file query was already rewritten, the directory query previously on line 815 could probably be modified too to include ``directory_fields`` directly in the sql query, instead of first fetching ids and then the fields using ORM? These changes would then have saved (at least) two trips to db.